doc/src/devdocs/ub.md

```diff
@@ -0,0 +1,50 @@
+
+*Undefined Behavior* (UB) occurs when a julia program semantically perform an operation that is assumed to never happen. In such a situation, the language semantics do not constrain the behavior of the implementation, so any behavior of the program is allowable, including crashes, memory corruption, incorrect behavior, etc. As such, it is very important to avoid writing programs that semantically execute undefined behavior.
+
+Note that this explicitly applies to *semantically executed* undefined behavior. While julia's compiler is allowed to and does aggressively perform speculative execution of pure functions. Since the execution point is not semantically observable (though again indirectly observable through execution performance), this is allowable by the as-if rule. As such, speculative execution is inhibited unless the code in question is proven to be
+free of undefined behavior.
+
+The presence of undefined behavior is modeled as part of julia's effect system using the `:noub` effect bit. See the documentation for `@assume_effects` for more information on querying the compiler's effect model or overriding it for specific situations (e.g. where a dynamic check precludes potential UB from every actually being reached).
+
+## List of sources of undefined behavior
+
+The following is a list of sources of undefined behavior,
+though it should currently not be considered exhaustive:
+
+- Replacement of `const` values. Note that in interactive mode the compiler will issue a warning for this and some care is taken to mimize impact as a user convenience, but the behavior is not defined.
+- Various modification of global state during precompile. Where possible, this is detected and yields and error, but detection is incomplete.
+- Incorrect implementation of a `Core.OptimizedGenerics` interface [1]
+- Any invocation of undefined behavior in FFI code (e.g. `ccall`, `llvmcall`) according to the semantics of the respective language
+- Incorrect signature types in `ccall` or `cfunction`, even if those signatures happen to yield the correct ABI on a particular platform
+- Incorrect use of annotations like `@inbounds`, `@assume_effects` in violation of their requirements [1]
+- Retention of pointers to GC-tracked objects outside of a `@GC.preserve` region
+- Memory modification of GC-tracked objects without appropriate write barriers from outside of julia (e.g. in native calls, debuggers, etc.)
+- Violations of the memory model using `unsafe` operations (e.g. `unsafe_load` of an invalid pointer, pointer provenance violations, etc)
+- Violations of TBAA guarantees (e.g. using `unsafe_wrap`)
```

Comment:
What does this mean? Is any use of `unsafe_wrap` enough for UB?

Reply:
If you unsafe_wrap a pointer from an array of one type in another, that's UB.

Reply:
Gotcha - I thought that's what you meant (since you mention TBAA), so it's probably good to clarify that in the docs. Maybe something like

```
- Violations of TBAA guarantees (e.g. by using `unsafe_wrap` to create a `Vector{Int}` from a pointer to a `Vector{UInt8}`)
```

would be good? This sort of warning should probably also be added to `unsafe_wrap` directly; `reinterpret` already warns of this, though more because of (in)valid bitpatterns than TBAA (which I think is mostly relevant for mutable types, if I'm not mistaken).

Reply:
sure

Reply:
> Is any use of unsafe_wrap enough for UB

That is a fairly hairy question. There is a significant amount of code in Base that does imply (via its documentation and compiler annotations) that any use of `unsafe_wrap` on any memory or pointer that already is accessible via another object will cause UB, not just if the type changes. But it is also not precisely the compiler that causes that UB, as is true of the rest of the items on the list. In this case, it is the runtime implementation instead which will violate its documented guarantees, possibly causing UB. Since we consider that documentation to be correct, therefore it must be UB for a user to use `unsafe_wrap` in such a manner.

Also, FWIW, we have moved away from this being merely a TBAA violation and towards being a total provenience violation across all of the multiple passes that may detect such occurrences. So perhaps 
```suggestion
- Violations of memory provenance guarantees (e.g. using `unsafe_wrap` on a pointer already accessible in the same scope or in the caller/callee of the same)
```

Reply:
Hmm.. Would that wording lead to it technically being UB when you use `unsafe_wrap` on a pointer retrieved through FFI? By necessity, the pointer must be accessible in the same scope before you can call `unsafe_wrap` - or am I misunderstanding what "accessible" means here?

Reply:
Access is about read or write ability, not the presence or value of the pointer

Reply:
So `unsafe_wrap` is only UB if I don't `unsafe_store!`/`unsafe_load` beforehand?

Reply:
`unsafe_load` and `unsafe_store` often (but not always) exist outside of the julia memory model (e.g. they are in the C memory model not the julia one), so combining them on the same object requires references the rules for doing those combinations